tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/service/websocket/RetryClientTest.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http:www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http:www.apache.orglicensesLICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specif ic language governing permissions and limitations
+ * under the License.
+ */
+
+package org.ballerinalang.test.service.websocket;
+
+import org.ballerinalang.test.context.BallerinaTestException;
+import org.ballerinalang.test.util.websocket.client.WebSocketTestClient;
+import org.ballerinalang.test.util.websocket.server.WebSocketRemoteServer;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.net.URISyntaxException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This Class tests the retry support of the WebSocket client and server.
+ */
+@Test(groups = {"websocket-test"})
+public class RetryClientTest extends WebSocketTestCommons {
+
+    private WebSocketRemoteServer remoteServer;
+    private static final String url =  "ws://localhost:21030";
+    private static final int port = 15300;
+    private static final String sentText = "hi all";
+
+    @Test(description = "Tests the retry function using the WebSocket client (starting the server " +
+            ", sending, and receiving text frames.  After that, restart that server and do the same.)")
+    public void testRetry() throws URISyntaxException, InterruptedException, BallerinaTestException {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        remoteServer = new WebSocketRemoteServer(port);
+        remoteServer.run();
+        WebSocketTestClient client = new WebSocketTestClient(url);
+        client.handshake();
+        client.setCountDownLatch(countDownLatch);
+        client.sendText(sentText);
+        countDownLatch.await(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
+        Assert.assertEquals(client.getTextReceived(), sentText);
+        remoteServer.stop();
+        CountDownLatch latchForRestart = new CountDownLatch(1);
+        client.setCountDownLatch(latchForRestart);
+        remoteServer.run();
+        latchForRestart.await(TIMEOUT_IN_SECS, TimeUnit.SECONDS);
```

Comment:
Why are you awaiting here? Can you read up on countdownlatches? We can await only once if we set countDown to 1. I think run is a blocking call. Can you please check other test cases and understand them? You can debug inside how remoteServer is implemented and then you will understand. The send* functions are non blocking and that is the reason we are using a countDown latch.

Reply:
When we don't use the countDownlatch, if remote server start, then client starts to send the data but here we can't do that. when a reconnect attempt, Client takes some time to connect with the server automatically after the server starts that depends on a retry's interval. Since we need to wait and send the data.

Reply:
Why are you setting it to the client then?

Reply:
Earlier, I didn't add countDownLatch to the client. I only added the counDownLatch of handshake in the client. After your comments(https://github.com/ballerina-platform/ballerina-lang/pull/19723#discussion_r350003437), I confused and added every countDownLatch to the client (https://github.com/ballerina-platform/ballerina-lang/pull/19723/commits/e79345f1bdbe83146ea7db96e49c8cf992648d43).

Yes, We don't need to set this countDown to the client. Now I have removed that changes.